docs/design/ha_master.md

```diff
@@ -0,0 +1,188 @@
+in front of apiservers we will have an L4 load balancer to spread the load among all the replicas.
+* Controller manager, scheduler & cluster autoscaler will use lease mechanism and
+only a single instance will be an active master. All other will be waiting in a standby mode.
+* All add-on managers will work independently and each of them will try to keep add-ons in sync
+
+# Detailed design
+
+## VM naming
+VMs with master replicas will have the following names:
+* first VM - `${MASTER_NAME}`
+* next VMs - `${MASTER_NAME}-<hash>`, where `<hash>` is a 3-characters long alphanumeric string.
+
+It will be ok to delete VM without the suffix.
+
+If user will run deployment scripts with a different `MASTER_NAME` variable than earlier,
+it will fail with an error that there are no masters to replicate.
+
+##Components
+
+###etcd
+```Note: this paragraphs assumes we are using etcd v2.2; it will have to be revisited during
+upgrade to v3, e.g. different command line options.```
+
+All etcd instances will be clustered together and one of them will be an elected master.
+In order to commit any change quorum of the cluster will have to confirm it. Etcd will be
+configured in such a way that all writes and reads will go through the master (requests
+will be forwarded by the local etcd server such that it’s invisible for the user). It will
+affect latency for all operations but it should not increase by much more than the network
+latency between master replicas (latency between zones with a region is < 10ms).
+
+Currently etcd exposes port only using localhost interface. In order to allow clustering
+and inter-VM communication we will also have to use public interface. To secure the
+communication we will configure firewall rules to only allow traffic on etcd port from
+other master machines.
+
+Another option to secure etcd cluster communication would be to use SSL (as described
+(here)[https://coreos.com/etcd/docs/latest/security.html]), but it would increase CPU usage
+on master machines, which is already a bottleneck in large clusters. As there is no clear
+benefits we decided to use a simpler and faster solution described above.
+
+When generating command line for etcd we will always assume it’s part of a cluster
+(initially of size 1) and list all VMs for kubernetes masters (see VM naming section).
+Based on that, we will set the following flags:
+* `-initial-cluster` - list of all hostnames for master machines
+* `-initial-cluster-state` (keep in mind that we are adding masters one by one):
+  * `new` if we are adding the first replica, i.e. the list of existing masters is empty
+  * `existing` if there are more than one replica, i.e. the list of existing masters is non-empty.
+
+This will allow us to have exactly the same logic for HA and non-HA master. List of VMs
+with master replicas will be generated in `kube-up.sh` script and passed to as a env variable
+`INITIAL_ETCD_CLUSTER`.
```

Comment:
I think you should call that out very explicitly, because now we have a DNS dependency in kube-up.  So... given we are using DNS, what else can we use DNS for?  Well, we have to use it for the LB on other clouds - are there advantages to using it on GCE also? I personally prefer to a cluster by name than by IP.  Should we use it internally rather than an internal LB etc.  Is it easier to use DNS to remap virtual names for etcd cluster members, rather than relying on the instance name - for example, will it simplify the case when # apiservers > # etcd nodes.